recursion/problem-set-recursion-pt-2.md

```diff
@@ -0,0 +1,290 @@
+    # besides the first item
+    return search(array[1:], needle)
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 2 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 0PjlSp
+* title: Recursion
+* points: 3
+### !question
+
+Write a recursive function `is_palindrome` that accepts a string `text` as a parameter. It returns a boolean value indicating if that string is a [palindrome](https://en.wikipedia.org/wiki/Palindrome) or not.
+
+| Input `text` | Return value |
+| ------------ | ------------ |
+| `"racecar"`  | `True`       |
+| `"raecar"`   | `False`      |
+
+Here are the tests:
+
+```python
+def test_is_palindrome_success():
+    assert is_palindrome("racecar")
+
+
+def test_is_palindrome_not_palindrome():
+    assert not is_palindrome("raecar")
+```
+
+### !end-question
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_is_palindrome_success(self):
+        self.assertTrue(is_palindrome("racecar"))
+
+    def test_is_palindrome_not_palindrome(self):
+        self.assertFalse(is_palindrome("raecar"))
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def is_palindrome(text):
+    if len(text) <= 1:
+        return True
+    # Use Python slicing to check if
+    # the first character is equal to
+    # the last character
+    elif text[:1] != text[-1:]:
+        return False
+    else:
+        # Use Python slicing to pass in
+        # the substring between the first
+        # and last characters of text
+        return is_palindrome(text[1:-1])
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+### !callout
+
+## Challenge for `is_palindrome`
+
+For an additional challenge, implement `is_palindrome` without creating new strings in the process of solving the problem.
+
+### !end-callout
+
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: lkW6SO
+* title: Recursion
+* points: 3
+### !question
+
+Write a recursive function named `digit_match`. It accepts two non-negative integers as parameters. It returns the number of digits that match in the two integers.
+
+Two digits match if they are equal _and_ have the same position relative to the end of the number (i.e. starting with the ones digit).
+
+In other words, the function should compare the last digits of each number, the second-to-last-digits-of-each number, the third-to-last digits of each number, and so fort, counting how many pairs match.
+
+Example:
+
+- First number: `1072503891`
+- Second number: `62530841`
+- Number of matches: `4`
+    - Matching pairs: 202, 5-5, 8-8, 1-1
+
+```
+1 0 7 2 5 0 3 8 9 1
+    | | | | | | | |
+    6 2 5 3 0 8 4 1
+```
+
+Here are the tests:
+
+```python
+def test_digit_match_large_inputs():
+    apples = 1072503891
+    oranges = 62530841
```

Comment:
apples and oranges is whimsical, and I like it, but like the earlier comments about needle (sans haystack) I wonder if this would be more confusing than not. Since these are just numbers, `a` and `b` might actually be clearer, but since the whole colocation is being used, I think I'm okay with sticking with apples and oranges. Just wanted to make sure there's some discussion around this.

Reply:
I love this discussion!

I've started using `apples` and `oranges` after getting a lot of feedback while teaching that `foo` `bar` `baz`, `a` `b` etc don't come across well, namely because they _look_ similar (length, literal shape)

Is there a place that you'd like to continue the convo?

Reply:
Oh, that's interesting. I guess I was indoctrinated into foo bar pretty early...  apples and oranges works well enough and there's a clear idiomatic reason to use them together. Given they look different as well, seems like a win!